webui/react/src/stores/experiments.tsx

```diff
@@ -1,63 +1,115 @@
-
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getExperiments(params, { signal: canceler.signal });
-      updateExperimentsCache(experimentsCache.set(encodeParams(params), response));
-    } catch (e) {
-      handleError(e);
+
+  // Get singleton instance
+  public static getInstance(): ExperimentsService {
+    if (!ExperimentsService.#instance) {
+      ExperimentsService.#isInternalConstructing = true;
+      ExperimentsService.#instance = new ExperimentsService();
+      ExperimentsService.#isInternalConstructing = false;
     }
-  }, [canceler, experimentsCache, params, updateExperimentsCache]);
-};
+    return ExperimentsService.#instance;
+  }
 
-export const useExperiments = (params: GetExperimentsParams): Loadable<ExperimentPagination> => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useExperiments outside of Experiment Context');
+  // Get an experiment by experiment id
+  public getExperimentsByIds(experimentIds: number[]): Observable<Loadable<ExperimentItem[]>> {
+    return this.#experimentMap.select((map) => {
+      const expList: ExperimentItem[] = [];
+      for (const id of experimentIds) {
+        const exp = map.get(id);
+        if (exp) {
+          expList.push(exp);
+        }
+      }
+      return Loaded(expList);
+    });
   }
-  const loadedVal = context.experimentsCache.get(encodeParams(params));
 
-  return loadedVal ? Loaded(loadedVal) : NotLoaded;
-};
+  public getExperimentsByParams(
+    params: Readonly<GetExperimentsParams>,
+  ): Observable<Loadable<ExperimentPagination>> {
+    return this.#experimentsCache.select((map) => {
+      const cache = map.get(encodeParams(params));
+      if (!cache) {
+        return NotLoaded;
+      }
+      const experiments: ExperimentItem[] = [];
+      const expMap = this.#experimentMap.get();
+      for (const id of cache.experimentIds) {
+        const expItem = expMap.get(id);
+        if (expItem) {
+          experiments.push(expItem);
+        }
+      }
+      const expPagination: ExperimentPagination = {
+        experiments: experiments,
+        pagination: cache.pagination,
+      };
+      return Loaded(expPagination);
+    });
+  }
+
+  // fetch experiments with params
+  public fetchExperiments(
+    params: Readonly<GetExperimentsParams>,
+    canceler: AbortController,
+  ): () => Promise<void> {
+    return async () => {
+      try {
+        const response = await getExperiments(params, { signal: canceler.signal });
+        this.#updateExperimentsCache(response, params);
+        this.#updateExperimentMap(response.experiments);
+      } catch (e) {
+        handleError(e);
+      }
+    };
+  }
+
+  #updateExperimentMap(experimentItems: Readonly<ExperimentItem[]>) {
+    this.#experimentMap.update((map) => {
+      let newMap: Map<number, ExperimentItem> = Map<number, ExperimentItem>();
```

Comment:
i wanna find a way to get rid of this let

Reply:
You can use withMutations here